File: node/Assembler/src/interface/ILoaderNormal.js
// ILoaderNormal interface (documentation + structural contract)
// Defines the minimal API that Normal engine relies on.

/**
 * @typedef {Object} ILoaderNormal
 * @property {string} searchAppSites - Comma-delimited list of fallback AppSites
 * @property {(rootDirPath:string, appSite:string, searchAppSites:string)=>boolean} load - Loads and caches templates
 * @property {(appSite:string, templateName:string)=>boolean} hasTemplate - Checks if a template exists
 * @property {()=>void} clearCache - Clears any internal caches
 * @property {(appSite:string, templateName:string, appView?:string|null, appViewPrefix?:string|null)=>string|null} getTemplateHtml - Gets raw HTML with AppView and fallback logic
 * @property {(html:string, appSite:string, templateName:string)=>string} mergeHtmlWithJson - Merges HTML with template JSON
 */

// This file intentionally exports a named symbol to make the interface discoverable in imports.
// Implementations should provide these methods; no runtime enforcement is performed.
export const ILoaderNormal = Symbol('ILoaderNormal');

File: rust/Assembler/src/interface/i_loader_normal.rs
/// Loader interface specifically for EngineNormal
/// Provides simple string-based template access without JSON support
/// Independent interface to ensure no coupling with other engine implementations
pub trait ILoaderNormal {
    /// Gets the search AppSites for template fallback resolution
    /// Comma-delimited string of AppSite names
    fn search_app_sites(&self) -> &str;

    /// Gets a template's HTML content by appSite and name with optional AppView fallback
    /// Returns raw HTML string only (no JSON merged)
    /// Searches in SearchAppSites if not found in primary appSite
    /// Templates stay immutable inside the loader
    ///
    /// # Arguments
    /// * `app_site` - The application site name
    /// * `template_name` - The template name (e.g., "Header", "Index")
    /// * `app_view` - Optional AppView for fallback logic
    /// * `app_view_prefix` - Optional AppView prefix for fallback logic
    ///
    /// # Returns
    /// Template HTML content or None if not found
    fn get_template_html(
        &self,
        app_site: &str,
        template_name: &str,
        app_view: Option<&str>,
        app_view_prefix: Option<&str>,
    ) -> Option<String>;

    /// Gets parsed JSON data for a template
    /// Returns None if no JSON file exists for the template
    /// Searches in SearchAppSites if not found in primary appSite
    ///
    /// # Arguments
    /// * `app_site` - The application site name
    /// * `template_name` - The template name
    ///
    /// # Returns
    /// Parsed JsonObject or None if no JSON file exists
    fn get_template_json(&self, app_site: &str, template_name: &str) -> Option<crate::app::json::JsonObject>;

    /// Merges HTML string with JSON data using loader-controlled inheritance logic
    /// Centralizes JSON merging so engines do not need to understand inheritance rules
    ///
    /// # Arguments
    /// * `html` - Raw HTML to merge
    /// * `app_site` - Application site for JSON lookup
    /// * `template_name` - Template name for JSON lookup
    ///
    /// # Returns
    /// HTML merged with JSON (or original HTML if no JSON is available)
    fn merge_html_with_json(&self, html: &str, app_site: &str, template_name: &str) -> String;

    /// Checks if a template exists
    ///
    /// # Arguments
    /// * `app_site` - The application site name
    /// * `template_name` - The template name
    ///
    /// # Returns
    /// True if template exists, false otherwise
    fn has_template(&self, app_site: &str, template_name: &str) -> bool;

    /// Clears the template cache (for testing/hot reload)
    fn clear_cache(&self);
}
